cirq/ops/two_qubit_diagonal_gate.py

```
@@ -0,0 +1,98 @@
+        self._diag_angles_radians: List[value.TParamVal] = diag_angles_radians
+
+    def _is_parameterized_(self):
+        return any(
+            isinstance(angle, sympy.Basic)
+            for angle in self._diag_angles_radians)
+
+    def _has_unitary_(self) -> bool:
+        return not self._is_parameterized_()
+
+    def _unitary_(self) -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        return np.diag(
+            [np.exp(1j * angle) for angle in self._diag_angles_radians])
+
+    def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs') -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        for index, angle in enumerate(self._diag_angles_radians):
+            subspace_index = args.subspace_index(big_endian_bits_int=index)
+            args.target_tensor[subspace_index] *= np.exp(1j * angle)
+        return args.target_tensor
+
+    def _resolve_parameters_(self, param_resolver: 'cirq.ParamResolver'
+                            ) -> 'TwoQubitDiagonalGate':
+        return TwoQubitDiagonalGate([
+            protocols.resolve_parameters(angle, param_resolver)
+            for angle in self._diag_angles_radians
+        ])
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'
+                              ) -> 'cirq.CircuitDiagramInfo':
+        rounded_angles = np.array(self._diag_angles_radians)
+        if args.precision is not None:
+            rounded_angles = rounded_angles.round(args.precision)
+        diag_str = 'diag({})'.format(', '.join(
+            proper_repr(angle) for angle in rounded_angles))
+        return protocols.CircuitDiagramInfo((diag_str, '#2'))
+
+    def __pow__(self, exponent: Any) -> 'TwoQubitDiagonalGate':
+        if not isinstance(exponent, (int, float, sympy.Basic)):
+            return NotImplemented
+        return TwoQubitDiagonalGate([
+            protocols.mul(angle, exponent, NotImplemented)
```

Why is the default `NotImplemented`.  Won't `TwoQubitDiagonalGate` break if given `NotImplemented` as an angle?  It may be better to either throw an error or return `NotImplemented` from `__pow__`.